rust/chains/hyperlane-cosmos/src/providers/mod.rs

```diff
@@ -172,14 +204,22 @@ impl HyperlaneProvider for CosmosProvider {
             .map_err(|e| ChainCommunicationError::from_other_str("could not convert from proto"))?;
         let contract = H256::try_from(CosmosAccountId::new(&msg.contract))?;
 
-        // TODO choose correct signer (should it be the one paid for the transaction)
-        let signer = tx
+        let (sender, nonce) = tx
             .auth_info
-            .signer_infos
-            .get(0)
-            .expect("there should be at least one signer");
-        let sender = self.sender(signer)?;
-        let nonce = signer.sequence;
+            .fee
+            .payer
+            .as_ref()
+            .map(|payer| self.search_payer_in_signer_infos(&tx.auth_info.signer_infos, payer))
+            .map_or_else(
```

Comment:
ditto I think `unwrap_or_else` could be clearer

Reply:
It is.